File: taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTarget.java
/**
 * Copyright 2016 Keepsafe Software, Inc.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.getkeepsafe.taptargetview;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Color;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.support.annotation.CallSuper;
import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.support.annotation.DimenRes;
import android.support.annotation.Dimension;
import android.support.annotation.DrawableRes;
import android.support.annotation.Nullable;
import android.support.annotation.Px;
import android.support.annotation.UiThread;
import android.support.v4.content.ContextCompat;

/**
 * Describes the properties and options for a {@link TapTargetView}.
 * <p>
 * Each tap target describes a target via a pair of bounds and icon. The bounds dictate the
 * location and touch area of the target, where the icon is what will be drawn within the center of
 * the bounds.
 * <p>
 * This class can be extended to support various target types.
 *
 * @see ViewTapTarget ViewTapTarget for targeting standard Android views
 */
public class TapTarget {
  private final Rect bounds;

  protected final Parameters param;

  private @Nullable TapTargetView parent;

  TapTarget(Parameters parameters) {
    // TODO: Ensure required params
    this.param = parameters;
    this.bounds = new Rect();
  }

  @UiThread
  public void setBounds(Rect bounds) {
    this.bounds.set(bounds);
    if (!isReady()) {
      return;
    }
    if (parent != null) {
      parent.onNewTargetBounds(bounds);
    }
  }

  @CallSuper
  @UiThread
  protected void attach(TapTargetView parent) {
    if (parent == null) {
      throw new IllegalArgumentException("Cannot attach to null parent");
    }
    if (this.parent != null) {
      detach();
    }
    this.parent = parent;
  }

  @CallSuper
  @UiThread
  protected void detach() {
    this.parent = null;
  }

  public @Nullable String id() {
    return param.id;
  }

  public boolean isReady() {
    return true;
  }

  protected static class TextParameters {
    public @Nullable CharSequence text;
    public Typeface typeface;
    public @ColorInt int color;
    public int size;
  }

  protected static class CircleParameters {
    public @ColorInt int color;
  }

  protected static class Parameters {
    public TextParameters title;
    public TextParameters description;

    public CircleParameters outerCircle;
    public CircleParameters targetCircle;

    public @Nullable String id;
    public @Nullable Drawable icon;

    public @ColorInt int dimColor;

    public boolean shadow;
    public boolean cancelable;
    public boolean tint;
    public boolean targetCircleTransparent;
  }

  public static class Builder {
    private final Context context;
    protected final Parameters parameters;

    Builder(Context context) {
      if (context == null) {
        throw new IllegalArgumentException("Given null Context");
      }
      this.context = context;

      // Setup defaults
      final boolean isDark = UiUtil.isDarkTheme(context);
      parameters = new Parameters();
      parameters.title = new TextParameters();
      parameters.title.color = isDark ? Color.BLACK : Color.WHITE;
      parameters.title.size = UiUtil.sp(context, 20);
      parameters.title.typeface = Typeface.create("sans-serif-medium", Typeface.NORMAL);

      parameters.description = new TextParameters();
      parameters.description.color = UiUtil.setAlpha(isDark ? Color.BLACK : Color.WHITE, 0.54f);
      parameters.description.size = UiUtil.sp(context, 18);
      parameters.description.typeface = Typeface.create(Typeface.SANS_SERIF, Typeface.NORMAL);

      final Resources.Theme theme = context.getTheme();
      final int outerCircleColor =
          (theme != null) ? UiUtil.themeIntAttr(context, "colorPrimary") : Color.WHITE;
      parameters.outerCircle = new CircleParameters();
      parameters.outerCircle.color = UiUtil.setAlpha(outerCircleColor, 0.96f);

      parameters.targetCircle= new CircleParameters();
      parameters.targetCircle.color = isDark ? Color.BLACK : Color.WHITE;

      parameters.dimColor = UiUtil.setAlpha(Color.BLACK, 0.3f);
      parameters.shadow = false;
      parameters.cancelable = true;
      parameters.tint = true;
      parameters.targetCircleTransparent = false;
    }

    /** Specify the color resource for the outer circle */
    public Builder outerCircleColorRes(@ColorRes int colorRes) {
      return outerCircleColor(ContextCompat.getColor(context, colorRes));
    }

    /** Specify the color value for the outer circle */
    public Builder outerCircleColor(@ColorInt int color) {
      parameters.outerCircle.color = color;
      return this;
    }

    /** Specify the color resource for the target circle */
    public Builder targetCircleColorRes(@ColorRes int colorRes) {
      return targetCircleColor(ContextCompat.getColor(context, colorRes));
    }

    /** Specify the color value for the target circle */
    public Builder targetCircleColor(@ColorInt int color) {
      parameters.targetCircle.color = color;
      return this;
    }

    /** Specify whether the target circle should be transparent */
    public Builder targetCircleIsTransparent(boolean status) {
      parameters.targetCircleTransparent = status;
      return this;
    }

    /** Specify the title text */
    public Builder titleText(CharSequence title) {
      parameters.title.text = title;
      return this;
    }

    /** Specify the color resource for the title text */
    public Builder titleTextColorRes(@ColorRes int colorRes) {
      return titleTextColor(ContextCompat.getColor(context, colorRes));
    }

    /** Specify the color value for the title text */
    public Builder titleTextColor(@ColorInt int color) {
      parameters.title.color = color;
      return this;
    }

    /** Specify the typeface for title text */
    public Builder titleTextTypeface(Typeface typeface) {
      checkNotNull(typeface, "title typeface");
      parameters.title.typeface = typeface;
      return this;
    }

    /** Specify the text size for the title from a dimension resource */
    public Builder titleTextSizeDimen(@DimenRes int dimen) {
      return titleTextSize(context.getResources().getDimensionPixelSize(dimen));
    }

    /** Specify the text size for the title in SP */
    public Builder titleTextSizeSp(@Dimension(unit = Dimension.SP) int sp) {
      return titleTextSize(UiUtil.sp(context, sp));
    }

    /** Specify the text size for the title in pixels */
    public Builder titleTextSize(@Px int px) {
      checkTextSize(px);
      parameters.title.size = px;
      return this;
    }

    /** Specify the description text */
    public Builder descriptionText(CharSequence description) {
      parameters.description.text = description;
      return this;
    }

    /** Specify the color resource for the description text */
    public Builder descriptionTextColorRes(@ColorRes int colorRes) {
      return descriptionTextColor(ContextCompat.getColor(context, colorRes));
    }

    /** Specify the color value for the description text */
    public Builder descriptionTextColor(@ColorInt int color) {
      parameters.description.color = color;
      return this;
    }

    /** Specify the typeface for description text */
    public Builder descriptionTextTypeface(Typeface typeface) {
      checkNotNull(typeface, "description typeface");
      parameters.description.typeface = typeface;
      return this;
    }

    /** Specify the text size for the description from a dimension resource */
    public Builder descriptionTextSizeDimen(@DimenRes int dimen) {
      return descriptionTextSize(context.getResources().getDimensionPixelSize(dimen));
    }

    /** Specify the text size for the description in SP */
    public Builder descriptionTextSizeSp(@Dimension(unit = Dimension.SP) int sp) {
      return descriptionTextSize(UiUtil.sp(context, sp));
    }

    /** Specify the text size for the description in pixels */
    public Builder descriptionTextSize(@Px int px) {
      checkTextSize(px);
      parameters.description.size = px;
      return this;
    }

    /** Specify the color resource to use as a dim effect */
    public Builder dimColorRes(@ColorRes int colorRes) {
      return dimColor(ContextCompat.getColor(context, colorRes));
    }

    /** Specify the color value to use as a dim effect */
    public Builder dimColor(@ColorInt int color) {
      parameters.dimColor = color;
      return this;
    }

    public Builder shadow(boolean status) {
      parameters.shadow = status;
      return this;
    }

    public Builder cancelable(boolean status) {
      parameters.cancelable = status;
      return this;
    }

    public Builder tintTarget(boolean status) {
      parameters.tint = status;
      return this;
    }

    /** Specify the icon resource that will be drawn in the center of the target bounds */
    public Builder iconRes(@DrawableRes int drawableRes) {
      return icon(ContextCompat.getDrawable(context, drawableRes));
    }

    /** Specify the icon that will be drawn in the center of the target bounds */
    public Builder icon(Drawable icon) {
      return icon(icon, false);
    }

    /**
     * Specify the icon that will be drawn in the center of the target bounds
     * @param hasSetBounds Whether the drawable already has its bounds correctly set. If the
     *                     drawable does not have its bounds set, then the following bounds will
     *                     be applied: <code>(0, 0, intrinsic-width, intrinsic-height)</code>
     */
    public Builder icon(Drawable icon, boolean hasSetBounds) {
      checkNotNull(icon, "icon");
      icon.mutate();
      if (!hasSetBounds) {
        icon.setBounds(new Rect(0, 0, icon.getIntrinsicWidth(), icon.getIntrinsicHeight()));
      }

      parameters.icon = icon;
      return this;
    }

    /** Specify a unique identifier for this target */
    public Builder id(String id) {
      parameters.id = id;
      return this;
    }

    void checkNotNull(Object object, String name) {
      if (object == null) throw new IllegalArgumentException("Given " + name + " is null");
    }

    void checkTextSize(int size) {
      if (size < 0) throw new IllegalArgumentException("Given negative text size");
    }

    public TapTarget build() {
      return new TapTarget(parameters);
    }
  }
}

File: taptargetview/src/main/java/com/getkeepsafe/taptargetview/TapTarget.kt
@file:Suppress("unused")
package com.getkeepsafe.taptargetview

import android.content.Context
import android.graphics.Rect
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.view.View
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.DimenRes
import androidx.annotation.IdRes
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat

open class TapTarget protected constructor(title: CharSequence?, description: CharSequence?) {
    val title: CharSequence
    val description: CharSequence?
    var outerCircleAlpha = 0.96f
    var targetRadius = 44
    var bounds: Rect? = null
    var icon: Drawable? = null
    var titleTypeface: Typeface? = null
    var descriptionTypeface: Typeface? = null

    @ColorRes
    private var outerCircleColorRes = -1

    @ColorRes
    private var targetCircleColorRes = -1

    @ColorRes
    private var dimColorRes = -1

    @ColorRes
    private var titleTextColorRes = -1

    @ColorRes
    private var descriptionTextColorRes = -1
    private var outerCircleColor: Int? = null
    private var targetCircleColor: Int? = null
    private var dimColor: Int? = null
    private var titleTextColor: Int? = null
    private var descriptionTextColor: Int? = null

    @DimenRes
    private var titleTextDimen = -1

    @DimenRes
    private var descriptionTextDimen = -1
    private var titleTextSize = 20
    private var descriptionTextSize = 18
    private var id = -1
    var drawShadow = false
    var cancelable = true
    var tintTarget = true
    var transparentTarget = false
    var descriptionTextAlpha = 0.54f

    protected constructor(
        bounds: Rect?,
        title: CharSequence?,
        description: CharSequence?
    ): this(title, description) {
        requireNotNull(bounds) { "Cannot pass null bounds or title" }
        this.bounds = bounds
    }

    init {
        requireNotNull(title) { "Cannot pass null title" }
        this.title = title
        this.description = description
    }

    /** Specify whether the target should be transparent  */
    fun transparentTarget(transparent: Boolean): TapTarget {
        transparentTarget = transparent
        return this
    }

    /** Specify the color resource for the outer circle  */
    fun outerCircleColor(@ColorRes color: Int): TapTarget {
        outerCircleColorRes = color
        return this
    }

    /** Specify the color value for the outer circle  */
    fun outerCircleColorInt(@ColorInt color: Int): TapTarget {
        outerCircleColor = color
        return this
    }

    /** Specify the alpha value [0.0, 1.0] of the outer circle  */
    fun outerCircleAlpha(alpha: Float): TapTarget {
        require(!(alpha < 0.0f || alpha > 1.0f)) { "Given an invalid alpha value: $alpha" }
        outerCircleAlpha = alpha
        return this
    }

    /** Specify the color resource for the target circle  */
    fun targetCircleColor(@ColorRes color: Int): TapTarget {
        targetCircleColorRes = color
        return this
    }

    /** Specify the color value for the target circle  */
    fun targetCircleColorInt(@ColorInt color: Int): TapTarget {
        targetCircleColor = color
        return this
    }

    /** Specify the color resource for all text  */
    fun textColor(@ColorRes color: Int): TapTarget {
        titleTextColorRes = color
        descriptionTextColorRes = color
        return this
    }

    /** Specify the color value for all text  */
    fun textColorInt(@ColorInt color: Int): TapTarget {
        titleTextColor = color
        descriptionTextColor = color
        return this
    }

    /** Specify the color resource for the title text  */
    fun titleTextColor(@ColorRes color: Int): TapTarget {
        titleTextColorRes = color
        return this
    }

    /** Specify the color value for the title text  */
    fun titleTextColorInt(@ColorInt color: Int): TapTarget {
        titleTextColor = color
        return this
    }

    /** Specify the color resource for the description text  */
    fun descriptionTextColor(@ColorRes color: Int): TapTarget {
        descriptionTextColorRes = color
        return this
    }

    /** Specify the color value for the description text  */
    fun descriptionTextColorInt(@ColorInt color: Int): TapTarget {
        descriptionTextColor = color
        return this
    }

    /** Specify the typeface for all text  */
    fun textTypeface(typeface: Typeface?): TapTarget {
        requireNotNull(typeface) { "Cannot use a null typeface" }
        titleTypeface = typeface
        descriptionTypeface = typeface
        return this
    }

    /** Specify the typeface for title text  */
    fun titleTypeface(titleTypeface: Typeface?): TapTarget {
        requireNotNull(titleTypeface) { "Cannot use a null typeface" }
        this.titleTypeface = titleTypeface
        return this
    }

    /** Specify the typeface for description text  */
    fun descriptionTypeface(descriptionTypeface: Typeface?): TapTarget {
        requireNotNull(descriptionTypeface) { "Cannot use a null typeface" }
        this.descriptionTypeface = descriptionTypeface
        return this
    }

    /** Specify the text size for the title in SP  */
    fun titleTextSize(sp: Int): TapTarget {
        require(sp >= 0) { "Given negative text size" }
        titleTextSize = sp
        return this
    }

    /** Specify the text size for the description in SP  */
    fun descriptionTextSize(sp: Int): TapTarget {
        require(sp >= 0) { "Given negative text size" }
        descriptionTextSize = sp
        return this
    }

    /**
     * Specify the text size for the title via a dimen resource
     *
     *
     * Note: If set, this value will take precedence over the specified sp size
     */
    fun titleTextDimen(@DimenRes dimen: Int): TapTarget {
        titleTextDimen = dimen
        return this
    }

    /** Specify the alpha value [0.0, 1.0] of the description text  */
    fun descriptionTextAlpha(descriptionTextAlpha: Float): TapTarget {
        require(!(descriptionTextAlpha < 0 || descriptionTextAlpha > 1f)) { "Given an invalid alpha value: $descriptionTextAlpha" }
        this.descriptionTextAlpha = descriptionTextAlpha
        return this
    }

    /**
     * Specify the text size for the description via a dimen resource
     *
     *
     * Note: If set, this value will take precedence over the specified sp size
     */
    fun descriptionTextDimen(@DimenRes dimen: Int): TapTarget {
        descriptionTextDimen = dimen
        return this
    }

    /**
     * Specify the color resource to use as a dim effect
     *
     *
     * **Note:** The given color will have its opacity modified to 30% automatically
     */
    fun dimColor(@ColorRes color: Int): TapTarget {
        dimColorRes = color
        return this
    }

    /**
     * Specify the color value to use as a dim effect
     *
     *
     * **Note:** The given color will have its opacity modified to 30% automatically
     */
    fun dimColorInt(@ColorInt color: Int): TapTarget {
        dimColor = color
        return this
    }

    /** Specify whether or not to draw a drop shadow around the outer circle  */
    fun drawShadow(draw: Boolean): TapTarget {
        drawShadow = draw
        return this
    }

    /** Specify whether or not the target should be cancelable  */
    fun cancelable(status: Boolean): TapTarget {
        cancelable = status
        return this
    }

    /** Specify whether to tint the target's icon with the outer circle's color  */
    fun tintTarget(tint: Boolean): TapTarget {
        tintTarget = tint
        return this
    }
    /**
     * Specify the icon that will be drawn in the center of the target bounds
     * @param hasSetBounds Whether the drawable already has its bounds correctly set. If the
     * drawable does not have its bounds set, then the following bounds will
     * be applied: <br></br>
     * `(0, 0, intrinsic-width, intrinsic-height)`
     */
    @JvmOverloads
    fun icon(icon: Drawable?, hasSetBounds: Boolean = false): TapTarget {
        requireNotNull(icon) { "Cannot use null drawable" }
        this.icon = icon
        if (!hasSetBounds) {
            icon.bounds = Rect(0, 0, icon.intrinsicWidth, icon.intrinsicHeight)
        }
        return this
    }

    /** Specify a unique identifier for this target.  */
    fun id(id: Int): TapTarget {
        this.id = id
        return this
    }

    /** Specify the target radius in dp.  */
    fun targetRadius(targetRadius: Int): TapTarget {
        this.targetRadius = targetRadius
        return this
    }

    /** Return the id associated with this tap target  */
    fun id(): Int {
        return id
    }

    /**
     * In case your target needs time to be ready (laid out in your view, not created, etc), the
     * runnable passed here will be invoked when the target is ready.
     */
    open fun onReady(runnable: Runnable?) {
        runnable?.run()
    }

    /**
     * Returns the target bounds. Throws an exception if they are not set
     * (target may not be ready)
     *
     *
     * This will only be called internally when [.onReady] invokes its runnable
     */
    fun bounds(): Rect {
        checkNotNull(bounds) { "Requesting bounds that are not set! Make sure your target is ready" }
        return bounds as Rect
    }

    fun outerCircleColorInt(context: Context): Int? {
        return colorResOrInt(context, outerCircleColor, outerCircleColorRes)
    }

    fun targetCircleColorInt(context: Context): Int? {
        return colorResOrInt(context, targetCircleColor, targetCircleColorRes)
    }

    fun dimColorInt(context: Context): Int? {
        return colorResOrInt(context, dimColor, dimColorRes)
    }

    fun titleTextColorInt(context: Context): Int? {
        return colorResOrInt(context, titleTextColor, titleTextColorRes)
    }

    fun descriptionTextColorInt(context: Context): Int? {
        return colorResOrInt(context, descriptionTextColor, descriptionTextColorRes)
    }

    fun titleTextSizePx(context: Context): Int {
        return dimenOrSize(context, titleTextSize, titleTextDimen)
    }

    fun descriptionTextSizePx(context: Context): Int {
        return dimenOrSize(context, descriptionTextSize, descriptionTextDimen)
    }

    private fun colorResOrInt(context: Context, value: Int?, @ColorRes resource: Int): Int? {
        return if (resource != -1) {
            ContextCompat.getColor(context, resource)
        } else value
    }

    private fun dimenOrSize(context: Context, size: Int, @DimenRes dimen: Int): Int {
        return if (dimen != -1) {
            context.resources.getDimensionPixelSize(dimen)
        } else size.sp
    }

    companion object {
        /** Return a tap target for the overflow button from the given toolbar
         *
         *
         * **Note:** This is currently experimental, use at your own risk
         */
        /**
         * Return a tap target for the overflow button from the given toolbar
         *
         *
         * **Note:** This is currently experimental, use at your own risk
         */
        @JvmOverloads
        fun forToolbarOverflow(
            toolbar: Toolbar?, title: CharSequence?,
            description: CharSequence? = null
        ): TapTarget {
            return ToolbarTapTarget(toolbar, false, title, description)
        }
        /** Return a tap target for the overflow button from the given toolbar
         *
         *
         * **Note:** This is currently experimental, use at your own risk
         */
        /** Return a tap target for the overflow button from the given toolbar
         *
         *
         * **Note:** This is currently experimental, use at your own risk
         */
        @JvmOverloads
        fun forToolbarOverflow(
            toolbar: android.widget.Toolbar?, title: CharSequence?,
            description: CharSequence? = null
        ): TapTarget {
            return ToolbarTapTarget(toolbar, false, title, description)
        }
        /** Return a tap target for the navigation button (back, up, etc) from the given toolbar  */
        /** Return a tap target for the navigation button (back, up, etc) from the given toolbar  */
        @JvmOverloads
        fun forToolbarNavigationIcon(
            toolbar: Toolbar?, title: CharSequence?,
            description: CharSequence? = null
        ): TapTarget {
            return ToolbarTapTarget(toolbar, true, title, description)
        }
        /** Return a tap target for the navigation button (back, up, etc) from the given toolbar  */
        /** Return a tap target for the navigation button (back, up, etc) from the given toolbar  */
        @JvmOverloads
        fun forToolbarNavigationIcon(
            toolbar: android.widget.Toolbar?, title: CharSequence?,
            description: CharSequence? = null
        ): TapTarget {
            return ToolbarTapTarget(toolbar, true, title, description)
        }
        /** Return a tap target for the menu item from the given toolbar  */
        /** Return a tap target for the menu item from the given toolbar  */
        @JvmOverloads
        fun forToolbarMenuItem(
            toolbar: Toolbar?, @IdRes menuItemId: Int,
            title: CharSequence?, description: CharSequence? = null
        ): TapTarget {
            return ToolbarTapTarget(toolbar, menuItemId, title, description)
        }
        /** Return a tap target for the menu item from the given toolbar  */
        /** Return a tap target for the menu item from the given toolbar  */
        @JvmOverloads
        fun forToolbarMenuItem(
            toolbar: android.widget.Toolbar?, @IdRes menuItemId: Int,
            title: CharSequence?, description: CharSequence? = null
        ): TapTarget {
            return ToolbarTapTarget(toolbar, menuItemId, title, description)
        }
        /** Return a tap target for the specified view  */
        /** Return a tap target for the specified view  */
        @JvmOverloads
        fun forView(
            view: View,
            title: CharSequence,
            description: CharSequence? = null
        ): TapTarget {
            return ViewTapTarget(view, title, description)
        }
        /** Return a tap target for the specified bounds  */
        /** Return a tap target for the specified bounds  */
        @JvmOverloads
        fun forBounds(
            bounds: Rect?,
            title: CharSequence?,
            description: CharSequence? = null
        ): TapTarget {
            return TapTarget(bounds, title, description)
        }
    }
}